File: Lab1/Lab1/T4.cs
using System.Threading;

namespace Lab1
{
    class T4
    {
        public static void Run(Data data)
        {
            data.print_message(Thread.CurrentThread.Name, "started");

            data.C = data.vector(data.isRandom);                    // 1
            data.x = data.number(data.isRandom);
            //data.print_vector("vector C: ", data.C);
            //data.print_number("number x: ", data.x);

            data.sem4.Release(3);                                   // 2-4
            data.sem1.WaitOne();                                    // 5
            data.sem2.WaitOne();                                    // 6
            data.sem3.WaitOne();                                    // 7

            data.mut1.WaitOne();
            int[] D = data.D;                                       // 8
            int[,] ME = data.ME;                                    // 9
            data.mut1.ReleaseMutex();

            data.operation1(D, ME, data.H * 3, data.H * 4);         // 10
            data.print_message(Thread.CurrentThread.Name, "operation 1 complete");

            data.evh3.Set();                                        // 11
            int ai = data.operation4(data.H * 3, data.H * 4);       // 12
            data.print_message(Thread.CurrentThread.Name, "operation 4 complete");

            lock (data._lock)
            {
                data.operation5(ai);                                // 13
            }
            data.print_message(Thread.CurrentThread.Name, "operation 5 complete");

            data.sem8.Release(3);                                   // 14-16
            data.sem5.WaitOne();                                    // 17
            data.sem6.WaitOne();                                    // 18
            data.sem7.WaitOne();                                    // 19

            data.mut2.WaitOne();
            int a = data.a;                                         // 20
            int x = data.x;                                         // 21
            data.mut2.ReleaseMutex();

            data.operation6(a, x, data.H * 3, data.H * 4);          // 22
            data.print_message(Thread.CurrentThread.Name, "operation 6 complete");
            data.bar1.SignalAndWait();                              // 23-25

            data.print_vector("result vector Z: ", data.Z);         // 26

        }
    }
}

File: Lab3/Lab3/src/T4.java
public class T4 extends Thread {
    private final Data data;

    public T4(Data data) {
        this.data = data;
    }

    public void run() {
        data.printMessage("T4", "started");

        data.M1.set_MT(Data.matrix());                              // 1
        data.M1.set_d(Data.number());

        data.M2.SignalIn();                                         // 2-4
        data.M2.WaitIn();                                           // 5-6

        int d4 = data.M1.get_d();                                   // 7
        int[] Z4 = data.M1.get_Z();                                 // 8

        data.M1.operation1(d4, Z4, data.H*3, data.H*4);             // 9
        data.printMessage("T4", "operation 1 complete");

        data.M3.SignalSort3();                                      // 10

        int a4 = data.M1.operation4(data.H*3, data.H*4);            // 11
        data.printMessage("T4", "operation 4 complete");

        data.M1.calc_a(a4);                                         // 12
        data.printMessage("T4", "operation 5 complete");

        data.M2.SignalA();                                          // 13-15
        data.M2.WaitA();                                            // 16-18

        int[] A4 = data.M1.get_A();                                 // 19
        int[][] MX4 = data.M1.get_MX();                             // 20
        a4 = data.M1.get_a();                                       // 21

        data.M1.operation6(A4, MX4, a4, data.H*3, data.H*4);        // 22
        data.printMessage("T4", "operation 6 complete");

        data.M2.SignalOut();                                        // 23-25
    }
}
